Upgrade to doctrine/dbal:4

johanib opened the pull request:
Needed for SF 6.4

MKodde commented on src/OpenConext/EngineBlock/Metadata/Entity/ServiceProvider.php:
```diff
@@ -48,19 +48,19 @@ class ServiceProvider extends AbstractRole
      * @var string[]
      */
-    #[ORM\Column(name: 'allowed_idp_entity_ids', type: \Doctrine\DBAL\Types\Types::ARRAY, length: 6777215)]
+    #[ORM\Column(name: 'allowed_idp_entity_ids', type: 'json', length: 16777215, nullable: true)]
```
Was the (significant) length increase intentional?

I also see that the field became nullable, but that is now explicitly set for all of the updated columns. I'm assuming that was previously the default setting which now needs an explicit setting.

johanib replied:
@MKodde I checked the length with: https://github.com/OpenConext/OpenConext-engineblock/issues/1861#issuecomment-3479676422

Its currently mediumtext on prod, so it should be 16777215. I htink 6777215 was a typo.

MKodde reviewed (commented):
Nice work! We had a small discussion IRL where I questioned the current solution:

1. Could you make a more 'generic' solution for the object and array types that you now added. This would need to include passing the type of the array items that are stored for additional type checking. And maybe passing a max length as some types have different requirements for that..
2. Keep the current setup but add some additional (unit or integration) tests. I've found a couple of unit tests [for two of the existing types](https://github.com/OpenConext/OpenConext-engineblock/blob/99fa3dd3a033250aab53c2a9c0818028de8d9d9f/tests/unit/OpenConext/EngineBlockBundle/Doctrine/Type/CollabPersonUuidTypeTest.php)

Question: are the types used in the existing functional tests? Or are they bypassed by the custom mock solution?

kayjoosten reviewed (changes requested):
Test are still failing

johanib commented:
> Nice work! We had a small discussion IRL where I questioned the current solution:
> 
> 1. Could you make a more 'generic' solution for the object and array types that you now added. This would need to include passing the type of the array items that are stored for additional type checking. And maybe passing a max length as some types have different requirements for that..
> 2. Keep the current setup but add some additional (unit or integration) tests. I've found a couple of unit tests [for two of the existing types](https://github.com/OpenConext/OpenConext-engineblock/blob/99fa3dd3a033250aab53c2a9c0818028de8d9d9f/tests/unit/OpenConext/EngineBlockBundle/Doctrine/Type/CollabPersonUuidTypeTest.php)
> 
> Question: are the types used in the existing functional tests? Or are they bypassed by the custom mock solution?

1. Yes, this was the correct way to go from the start. Implemented now. 
2. Also added tests for the two 'new' types. 
3. The behat tests fail if I omit the serialize, so yes, they are used in the behat tests.

MKodde commented on src/OpenConext/EngineBlock/Metadata/Entity/AbstractRole.php:
```diff
@@ -134,25 +136,25 @@ abstract class AbstractRole
     /**
      * @var Organization
      */
-    #[ORM\Column(name: 'organization_nl_name', type: \Doctrine\DBAL\Types\Types::OBJECT, nullable: true, length: 65535)]
```
Let me start by complimenting you on this change. This makes the PR so much more to the point!

Two things caught my eye: 
- Is there no need to set the column length anymore? Did we set the default value excessively before?
- And, I was also triggered by you setting the `nullable` value for each field now. I'm OK with that, before we would only set it if the non-default value was required. 
- Does this change cause 0 schema changes? If it does, you might need to think about a strategy to roll out that change safely. But I guess there are none judging the code.

MKodde reviewed (approved):
Nice work! Thanks for taking extra care in checking the schema changes, `required` and `length` constraints.
